### PR TITLE
fix: handle empty hands in resolve_round (issue #144)

### LINKED ISSUE #144: [Fix][Doctor Needed] Adjust `resolve_game` in `InternalTrait`
## Overview
When resolving a game, for player in round, if the `Hand` model is read, the cards in the hand should be checked. if empty, just skip this player. It just means the player didn't submit a hand. For now, it's a big sorry to this player 😔.

Anyways, In the end, this function should reset all player's stored cards in the `hand` the world to an empty array, or probably just delete the hand. Then also reset both roots in the game to zero.

## Furthermore?
- The target branch is the v1 branch.
- Please when applying, state whether you’re new to the whole language and all. If you are, I could help you steadily (Though it’ll stress me out)… if you aren’t, better for me… I’ll leave everything in your hands 😌, and you won’t be permitted to paste errors in my dm.
- Please use all available traits possible at your disposal 
- Your code must be **strongly typed** for better readability please 🙏🏻
- Document your code please, in a professional manner. Please avoid using too many small comments as you should know, only comment on hard to understand logic, and maybe some explanations on how and why it works
- When done, @ your username in a comment on top of the function you touched. If there's one there already, append with a comma and a space. (`// @ExistingUsername, @YourUsername`, etc)... It's cooler to leave your signature behind, yes? 😋
- Each `//` should be accompanied by a space, before typing, lock in.
- Please avoid code repitition. There may be lines of codes that would be reused, please extract them into a separate function. Lock in your good PR quality please 🙏🏻
- May the force be with you ☪️✋🏻
- Ofcourse, if you have any questions (Which I bet you'd have, one billion per cent), you can branch [here](https://t.me/lemonade46) or [here](https://t.me/+tqBpITsr5mllZDQ0). Happy Hacks 😊💙

## Help?
If you want it, you can have it. Find it! I left everything this world has to offer _[there](https://t.me/+tqBpITsr5mllZDQ0)_ -- _**Pirate Gol D. Roger**_

### PR DESCRIPTION
**Pull request by NathaliaBarreiros**
## Description
This PR implements the requirements for issue [#144](https://github.com/Prometheus-A/The-Velvet-Ace/issues/144):  
- When resolving a round, players with empty hands are skipped.
- All player hands are cleaned at the end of the round.
- Both `deck_root` and `dealt_cards_root` are reset to zero after each round.
- The game and player states are updated and prepared for the next round.

## Related Issues
Fixes [#144](https://github.com/Prometheus-A/The-Velvet-Ace/issues/144)

## Changes Made
- [x] Added `resolve_round` to the contract interface (`fn resolve_round(ref self: TContractState, game_id: u64);`) so it can be called from tests and external contracts, enabling proper testing and usage of the round resolution logic.
- [x] Registered the `RoundResolved` event in the test setup (`TestResource::Event(e_RoundResolved::TEST_CLASS_HASH),`) to allow event emission and verification during tests.
- [x] Implemented `_resolve_round` logic in `actions.cairo` to:
  - Skip players with empty hands.
  - Clean all player hands and reset roots.
  - Update game and player states for the next round.
- [x] Created `_cleanup_game_data` function for clarity and reuse.
- [x] Added tests in `test_resolve_round.cairo` to cover all scenarios, including:
  - Skipping empty hands
  - Cleaning all cards
  - Resetting roots
  - Updating game and player states
  - Handling edge cases

## How to Test
1. Run all tests in the contract directory:
   ```bash
   sozo test
   ```
2. All tests in `test_resolve_round.cairo` should pass, covering the new logic for round resolution and cleanup.

## Screenshots (if applicable)
N/A

## Checklist
- [x] My code follows the project's coding style.
- [x] I have tested these changes locally.
- [x] Documentation has been updated where necessary.

### CODE REVIEW

**Comment by NathaliaBarreiros** [comment]
@Birdmannn, I had some issues with the versions while setting up the project, which is why I changed them. But it seems to be failing the checks' review. Gonna change this. Lmk if something else needs to be changed.

**Comment by Birdmannn** [comment]
@NathaliaBarreiros update the workflows to reflect the version you've changed then. Then I can start a review when I'm sure the workflow passes and the code is running well. Thank you very much 🙏🏻

**Comment by Birdmannn** [comment]
@NathaliaBarreiros `scarb fmt` please. I'll review in a few minutes

File: poker-texas-hold-em/contract/manifest_dev.json [hunk 1]
```diff
@@ -1252,8 +1328,8 @@
   },
   "contracts": [
     {
-      "address": "0x2115aaa49adb5178478e04d8707ab7fe361aa9e140866b7365c3bb1f1d93334",
-      "class_hash": "0x6cbb5ab40ad7b48dff70edeacc4902a9eeae63a9e36456f9f908aedb904fbd2",
+      "address": "0xe234d63a9e7355bf8ff01d33111ac2852ecc38c6ae7ef00d35d0cd13979a5e",
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/manifest_dev.json]: Did you deploy this contract? If yes, why? Though I see no need to deploy the contract. If there was no tangible reason, sorry miss, you might have to revert 👀.

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 2]
```diff
@@ -707,89 +711,120 @@ pub mod actions {
             world.emit_event(@HandResolved { game_id: game_id, players: resolved_players });
         }
 
-        /// @psychemist, @Birdmannn
+        /// @psychemist, @Birdmannn, @NathaliaBarreiros
         ///
-        /// Resolves the current round and prepares the game for the next round
+        /// Resolves the current round and prepares the game for the next round.
+        /// Implements issue #144: skips players with empty hands and resets game roots.
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo:717]: Don't delete the previous comments. You can just combine both, but this line should be exempted

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 3]
```diff
@@ -335,6 +335,10 @@ pub mod actions {
 
             world.write_model(@player);
         }
+
+        fn resolve_round(ref self: ContractState, game_id: u64) {
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo:339]: fairs

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 4]
```diff
@@ -707,89 +711,120 @@ pub mod actions {
...
+            
+            // Clean up all player hands
+            let mut i: u32 = 0;
+            while i < players.len() {
+                let player_address = *players.at(i);
+                let mut hand: Hand = world.read_model(player_address);
+                hand.new_hand(); // Clear all cards from this player's hand
+                world.write_model(@hand);
+                i += 1;
+            };
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo:827]: I thought this has been cleaned by you before, using `self._resolve_hands()`?

File: poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+        let hand_2 = create_hand_with_cards(
+            PLAYER_2(),
+            array![create_card(12, Suits::CLUBS), create_card(11, Suits::DIAMONDS)]
+        );
+        let hand_3 = create_hand_with_cards(
+            PLAYER_3(),
+            array![create_card(10, Suits::SPADES), create_card(9, Suits::HEARTS)]
+        );
+        
+        world.write_models(array![@hand_1, @hand_2, @hand_3].span());
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo]: here

File: poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo]: here

File: poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+        let hand_2 = create_hand_with_cards(
+            PLAYER_2(),
+            array![create_card(12, Suits::CLUBS), create_card(11, Suits::DIAMONDS)]
+        );
+        let hand_3 = create_hand_with_cards(
+            PLAYER_3(),
+            array![create_card(10, Suits::SPADES), create_card(9, Suits::HEARTS)]
+        );
+        
+        world.write_models(array![@hand_1, @hand_2, @hand_3].span());
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo]: here

File: poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo]: here

File: poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+        assert(player_1_after.current_bet == 0, 'P1 current_bet should be reset');
+        assert(player_2_after.current_bet == 0, 'P2 current_bet should be reset');
+        assert(player_3_after.current_bet == 0, 'P3 current_bet should be reset');
+        
+        assert(player_1_after.in_round == true, 'P1 should be in next round');
+        assert(player_2_after.in_round == true, 'P2 should be in next round');
+        assert(player_3_after.in_round == true, 'P3 should be in next round');
+    }
+
+    /// Test mixed scenario: some players with cards, some with empty hands
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo]: This test has been carried out before, but in a different way

File: poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+        assert(hand_3_after.cards.len() == 0, 'P3 hand should be cleaned');
+        
+        // Game should still progress normally
+        let game: Game = world.read_model(1_u64);
+        assert(!game.round_in_progress, 'Round should not be in progress');
+        assert(game.deck_root == 0, 'Deck root should be reset');
+        assert(game.dealt_cards_root == 0, 'Dealt root should be reset');
+    }
+
+    /// Test community cards are cleared after round resolution
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo:290]: Did you know that all these tests/checks can be written in just one test? Say, a maximum of two or three, if you really want to avoid bulk code blocks. There's quite an overhead in re-writing things you've written in the past just to assert 3 lines. There are just too many repeated code, but at what cost? Yes, I always understand that each panic test should be separated, because they panic... But if truly you wish to separate test cases for each particular test (which I don't see a need to, unless the tests use a long flow), then you just create helper functions that hold repeated code... then you can just call the helper functions in each one of the test, and then all that's left is to make the necessary assertions.

Trust me, the more concise the code is, the better. I don't think anybody likes long code... well maybe you and a few others 👀. 

Now that you have written 600+ lines of code however you wrote it, it'll take you an overhead to compress these tests.

**Review by Birdmannn — COMMENTED** [review]
Seems you'll have to revert the manifest. Most of the time used to refactor things and comments out of your task would've been useful and put in refactoring your tests to a loveable one, miss 🥲.

So what's your proposal?

**Comment by Birdmannn** [comment]
@NathaliaBarreiros

File: poker-texas-hold-em/contract/src/systems/actions.cairo (hunk 4 above)
Comment (NathaliaBarreiros) [reply poker-texas-hold-em/contract/src/systems/actions.cairo:827]: You're correct about the duplication. I've refactored the code:

- Removed: `self._resolve_hands(ref valid_players)`, which only cleaned valid players.
-Kept: `_cleanup_game_data()` which cleans ALL players as required by the issue

This eliminates the duplication and ensures all player hands are properly reset, not just the ones with cards. Much cleaner now!

Comment (Birdmannn) [reply poker-texas-hold-em/contract/src/systems/actions.cairo:827]: Yosh 💯

File: poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo (hunk 10 above)
Comment (NathaliaBarreiros) [reply poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo:290]: You're right! I've already refactored the tests and created helper functions to eliminate the repetitive code:
- `setup_test_environment()` - centralized setup
- `create_standard_hands()` - reusable hand creation
- `assert_all_hands_cleaned()` &`assert_game_state_after_resolution()` - common assertions

thnx for the feedback!

File: poker-texas-hold-em/contract/manifest_dev.json (hunk 1 above)
Comment (NathaliaBarreiros) [reply poker-texas-hold-em/contract/manifest_dev.json]: I accidentally deployed the contract while testing the project and forgot to revert this change. I reverted the manifest back to the original state.

Comment (Birdmannn) [reply poker-texas-hold-em/contract/manifest_dev.json]: Wow

File: poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo (hunk 9 above)
Comment (NathaliaBarreiros) [reply poker-texas-hold-em/contract/src/tests/test_resolve_round.cairo]: I've consolidated tests into a cleaner version, deleting duplicate test cases.

File: poker-texas-hold-em/contract/Scarb.toml [hunk 11]
```diff
@@ -1,5 +1,5 @@
 [package]
-cairo-version = "=2.10.1"
+cairo-version = "=2.9.2"
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/Scarb.toml]: You chose the `incoming` Scarb.toml?
 
You have two choices here, to either choose the cairo 2.9.2 you fetched from upstream, and rework the earlier committed workflow adjustments you made, 

Or


Match this scarb.toml to your committed workflow

**Review by Birdmannn — COMMENTED** [review]
Check

File: poker-texas-hold-em/contract/Scarb.toml (hunk 11 above)
Comment (NathaliaBarreiros) [reply poker-texas-hold-em/contract/Scarb.toml]: Done!

**Comment by Birdmannn** [comment]
Good job @NathaliaBarreiros ❤️‍🔥 You're the kind of dev this world needs. You very attentive to corrections and the details the task needs. You do things extensively, and most importantly, you do them well. Maybe in the future, you might try out hard tasks with a more complex logic 🧘🏻‍♂️. I hope our paths cross again. Muchas graçias 🙇🏻♠️
